File: test-application.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"net/http"
	"time"
)

// User represents a user in our system
type User struct {
	ID   int    `json:"id"`
	Name string `json:"name"`
}

// Response represents a standard API response
type Response struct {
	Status  string      `json:"status"`
	Message string      `json:"message"`
	Data    interface{} `json:"data,omitempty"`
}

// simulateDatabase simulates a database call with random latency
func simulateDatabase(ctx context.Context, operation string) error {
	// Random latency between 10-100ms to make traces interesting
	latency := time.Duration(rand.Intn(90)+10) * time.Millisecond
	log.Printf("Database operation: %s (latency: %v)", operation, latency)

	select {
	case <-time.After(latency):
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}

// simulateExternalAPI simulates calling an external API
func simulateExternalAPI(ctx context.Context, endpoint string) (interface{}, error) {
	log.Printf("Calling external API: %s", endpoint)

	// Simulate network latency
	latency := time.Duration(rand.Intn(200)+50) * time.Millisecond

	select {
	case <-time.After(latency):
		// Simulate some data being returned
		return map[string]interface{}{
			"external_data": fmt.Sprintf("Data from %s", endpoint),
			"timestamp":     time.Now().Unix(),
		}, nil
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

// goodHandler handles the /good endpoint
func goodHandler(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	log.Printf("Processing good request from %s", r.RemoteAddr)

	// Simulate some business logic with database calls
	if err := simulateDatabase(ctx, "SELECT users WHERE active=true"); err != nil {
		log.Printf("Database error: %v", err)
		http.Error(w, "Database unavailable", http.StatusServiceUnavailable)
		return
	}

	// Simulate calling an external service
	externalData, err := simulateExternalAPI(ctx, "https://api.example.com/status")
	if err != nil {
		log.Printf("External API error: %v", err)
		// Continue anyway for demo purposes
	}

	// Create some users data
	users := []User{
		{ID: 1, Name: "Alice Johnson"},
		{ID: 2, Name: "Bob Smith"},
		{ID: 3, Name: "Charlie Brown"},
	}

	response := Response{
		Status:  "success",
		Message: "Request processed successfully",
		Data: map[string]interface{}{
			"users":         users,
			"external_data": externalData,
			"processed_at":  time.Now().Format(time.RFC3339),
		},
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)

	if err := json.NewEncoder(w).Encode(response); err != nil {
		log.Printf("Error encoding response: %v", err)
	}

	log.Printf("Successfully processed good request")
}

// badHandler handles the /bad endpoint
func badHandler(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	log.Printf("Processing bad request from %s", r.RemoteAddr)

	// Simulate some processing that leads to an error
	if err := simulateDatabase(ctx, "SELECT * FROM non_existent_table"); err != nil {
		log.Printf("Expected database error: %v", err)
	}

	// Simulate multiple failed operations
	operations := []string{
		"validate_user_permissions",
		"check_rate_limits",
		"process_payment",
	}

	for _, op := range operations {
		log.Printf("Operation failed: %s", op)
		// Add some artificial delay to make traces more interesting
		time.Sleep(time.Duration(rand.Intn(20)+5) * time.Millisecond)
	}

	// Try external API call that will "fail"
	_, err := simulateExternalAPI(ctx, "https://api.example.com/broken-endpoint")
	if err != nil {
		log.Printf("External API call failed as expected: %v", err)
	}

	response := Response{
		Status:  "error",
		Message: "Internal server error occurred",
		Data: map[string]interface{}{
			"error_code": "INTERNAL_ERROR",
			"timestamp":  time.Now().Format(time.RFC3339),
		},
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusInternalServerError)

	if err := json.NewEncoder(w).Encode(response); err != nil {
		log.Printf("Error encoding error response: %v", err)
	}

	log.Printf("Processed bad request with error response")
}

// adminHandler handles the /admin endpoint
func adminHandler(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	log.Printf("Admin access attempted from %s", r.RemoteAddr)

	// Simulate authentication check
	authToken := r.Header.Get("Authorization")
	log.Printf("Checking authorization token: %s", authToken)

	// Simulate database call to check permissions
	if err := simulateDatabase(ctx, "SELECT permissions FROM users WHERE token=?"); err != nil {
		log.Printf("Auth database error: %v", err)
	}

	// Simulate permission validation logic
	operations := []string{
		"validate_token_format",
		"check_token_expiry",
		"verify_admin_permissions",
	}

	for _, op := range operations {
		log.Printf("Auth operation: %s", op)
		time.Sleep(time.Duration(rand.Intn(15)+5) * time.Millisecond)
	}

	// Always return unauthorized for demo purposes
	log.Printf("Authorization failed - insufficient permissions")

	response := Response{
		Status:  "error",
		Message: "Unauthorized access - admin privileges required",
		Data: map[string]interface{}{
			"error_code":    "UNAUTHORIZED",
			"required_role": "admin",
			"timestamp":     time.Now().Format(time.RFC3339),
		},
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusUnauthorized)

	if err := json.NewEncoder(w).Encode(response); err != nil {
		log.Printf("Error encoding unauthorized response: %v", err)
	}

	log.Printf("Rejected admin request - unauthorized")
}

// healthHandler provides a simple health check
func healthHandler(w http.ResponseWriter, r *http.Request) {
	log.Printf("Health check from %s", r.RemoteAddr)

	response := Response{
		Status:  "healthy",
		Message: "Service is running",
		Data: map[string]interface{}{
			"uptime":    time.Since(startTime).String(),
			"timestamp": time.Now().Format(time.RFC3339),
		},
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(response)
}

var startTime time.Time

func main() {
	startTime = time.Now()

	// Seed random number generator for consistent but varied latencies
	rand.Seed(time.Now().UnixNano())

	// Set up routes
	http.HandleFunc("/good", goodHandler)
	http.HandleFunc("/bad", badHandler)
	http.HandleFunc("/admin", adminHandler)
	http.HandleFunc("/health", healthHandler)

	// Root handler
	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/" {
			http.NotFound(w, r)
			return
		}

		response := Response{
			Status:  "success",
			Message: "OpenTelemetry Demo Server",
			Data: map[string]interface{}{
				"endpoints": []string{"/good", "/bad", "/admin", "/health"},
				"version":   "1.0.0",
			},
		}

		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(response)
	})

	port := ":8080"
	log.Printf("Starting server on port %s", port)
	log.Printf("Available endpoints:")
	log.Printf("  GET /        - Service info")
	log.Printf("  GET /good    - Returns 200 with success response")
	log.Printf("  GET /bad     - Returns 500 with error response")
	log.Printf("  GET /admin   - Returns 401 unauthorized")
	log.Printf("  GET /health  - Health check endpoint")

	if err := http.ListenAndServe(port, nil); err != nil {
		log.Fatalf("Server failed to start: %v", err)
	}
}

File: test-application.py
#!/usr/bin/env python3
"""
OpenTelemetry Demo Server - Python Version
A simple HTTP server that generates various types of traces for testing
"""

import json
import logging
import random
import time
from datetime import datetime
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import Dict, Any

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

# Global start time for uptime calculation
start_time = datetime.now()


class User:
    """Represents a user in our system"""
    def __init__(self, id: int, name: str):
        self.id = id
        self.name = name
    
    def to_dict(self):
        return {"id": self.id, "name": self.name}


def simulate_database(operation: str) -> None:
    """Simulates a database call with random latency"""
    # Random latency between 10-100ms to make traces interesting
    latency = random.randint(10, 100) / 1000  # Convert to seconds
    logger.info(f"Database operation: {operation} (latency: {latency*1000:.0f}ms)")
    time.sleep(latency)


def simulate_external_api(endpoint: str) -> Dict[str, Any]:
    """Simulates calling an external API"""
    logger.info(f"Calling external API: {endpoint}")
    
    # Simulate network latency
    latency = random.randint(50, 250) / 1000  # Convert to seconds
    time.sleep(latency)
    
    # Simulate some data being returned
    return {
        "external_data": f"Data from {endpoint}",
        "timestamp": int(time.time())
    }


class RequestHandler(BaseHTTPRequestHandler):
    """HTTP request handler for the demo server"""
    
    def log_message(self, format, *args):
        """Override to use our logger"""
        logger.info(f"{self.address_string()} - {format % args}")
    
    def send_json_response(self, status_code: int, response: Dict[str, Any]):
        """Helper to send JSON responses"""
        self.send_response(status_code)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(json.dumps(response).encode('utf-8'))
    
    def do_GET(self):
        """Handle GET requests"""
        if self.path == '/':
            self.handle_root()
        elif self.path == '/good':
            self.handle_good()
        elif self.path == '/bad':
            self.handle_bad()
        elif self.path == '/admin':
            self.handle_admin()
        elif self.path == '/health':
            self.handle_health()
        else:
            self.send_error(404, "Not Found")
    
    def handle_root(self):
        """Root handler - returns service info"""
        response = {
            "status": "success",
            "message": "OpenTelemetry Demo Server (Python)",
            "data": {
                "endpoints": ["/good", "/bad", "/admin", "/health"],
                "version": "1.0.0"
            }
        }
        self.send_json_response(200, response)
    
    def handle_good(self):
        """Handle /good endpoint - successful request with various operations"""
        logger.info(f"Processing good request from {self.address_string()}")
        
        try:
            # Simulate some business logic with database calls
            simulate_database("SELECT users WHERE active=true")
            
            # Simulate calling an external service
            external_data = None
            try:
                external_data = simulate_external_api("https://api.example.com/status")
            except Exception as e:
                logger.error(f"External API error: {e}")
                # Continue anyway for demo purposes
            
            # Create some users data
            users = [
                User(1, "Alice Johnson"),
                User(2, "Bob Smith"),
                User(3, "Charlie Brown")
            ]
            
            response = {
                "status": "success",
                "message": "Request processed successfully",
                "data": {
                    "users": [user.to_dict() for user in users],
                    "external_data": external_data,
                    "processed_at": datetime.now().isoformat()
                }
            }
            
            self.send_json_response(200, response)
            logger.info("Successfully processed good request")
            
        except Exception as e:
            logger.error(f"Error processing good request: {e}")
            self.send_error(503, "Service Unavailable")
    
    def handle_bad(self):
        """Handle /bad endpoint - returns 500 error after some processing"""
        logger.info(f"Processing bad request from {self.address_string()}")
        
        # Simulate some processing that leads to an error
        try:
            simulate_database("SELECT * FROM non_existent_table")
        except Exception as e:
            logger.info(f"Expected database error: {e}")
        
        # Simulate multiple failed operations
        operations = [
            "validate_user_permissions",
            "check_rate_limits",
            "process_payment"
        ]
        
        for op in operations:
            logger.info(f"Operation failed: {op}")
            # Add some artificial delay to make traces more interesting
            time.sleep(random.randint(5, 25) / 1000)
        
        # Try external API call that will "fail"
        try:
            simulate_external_api("https://api.example.com/broken-endpoint")
        except Exception as e:
            logger.info(f"External API call failed as expected: {e}")
        
        response = {
            "status": "error",
            "message": "Internal server error occurred",
            "data": {
                "error_code": "INTERNAL_ERROR",
                "timestamp": datetime.now().isoformat()
            }
        }
        
        self.send_json_response(500, response)
        logger.info("Processed bad request with error response")
    
    def handle_admin(self):
        """Handle /admin endpoint - returns 401 unauthorized"""
        logger.info(f"Admin access attempted from {self.address_string()}")
        
        # Simulate authentication check
        auth_token = self.headers.get('Authorization', '')
        logger.info(f"Checking authorization token: {auth_token}")
        
        # Simulate database call to check permissions
        try:
            simulate_database("SELECT permissions FROM users WHERE token=?")
        except Exception as e:
            logger.error(f"Auth database error: {e}")
        
        # Simulate permission validation logic
        operations = [
            "validate_token_format",
            "check_token_expiry",
            "verify_admin_permissions"
        ]
        
        for op in operations:
            logger.info(f"Auth operation: {op}")
            time.sleep(random.randint(5, 20) / 1000)
        
        # Always return unauthorized for demo purposes
        logger.info("Authorization failed - insufficient permissions")
        
        response = {
            "status": "error",
            "message": "Unauthorized access - admin privileges required",
            "data": {
                "error_code": "UNAUTHORIZED",
                "required_role": "admin",
                "timestamp": datetime.now().isoformat()
            }
        }
        
        self.send_json_response(401, response)
        logger.info("Rejected admin request - unauthorized")
    
    def handle_health(self):
        """Handle /health endpoint - health check"""
        logger.info(f"Health check from {self.address_string()}")
        
        uptime = datetime.now() - start_time
        response = {
            "status": "healthy",
            "message": "Service is running",
            "data": {
                "uptime": str(uptime),
                "timestamp": datetime.now().isoformat()
            }
        }
        
        self.send_json_response(200, response)


def run_server(port: int = 8080):
    """Run the HTTP server"""
    server_address = ('', port)
    httpd = HTTPServer(server_address, RequestHandler)
    
    logger.info(f"Starting server on port {port}")
    logger.info("Available endpoints:")
    logger.info("  GET /        - Service info")
    logger.info("  GET /good    - Returns 200 with success response")
    logger.info("  GET /bad     - Returns 500 with error response")
    logger.info("  GET /admin   - Returns 401 unauthorized")
    logger.info("  GET /health  - Health check endpoint")
    
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        logger.info("Server shutting down...")
        httpd.shutdown()


if __name__ == "__main__":
    # Seed random number generator for consistent but varied latencies
    random.seed()
    
    # Run the server
    run_server()